toolchain/check/interface.cpp

```diff
@@ -241,18 +242,50 @@ auto GetSelfParameter(Context& context, SemIR::TypeId type_id,
   return self_param_inst_id;
 }
 
-auto TryGetExistingDecl(
-    Context& context, SemIR::LocId loc_id, const NameComponent& name,
-    const DeclNameStack::NameContext& name_context,
-    Lex::TokenKind decl_token_kind, const SemIR::EntityWithParamsBase& entity,
-    bool is_definition,
-    llvm::function_ref<auto(SemIR::Inst)->const SemIR::EntityWithParamsBase*>
-        try_get_entity,
-    SemIR::ScopeLookupResult lookup_result) -> std::optional<SemIR::Inst> {
+template <typename EntityT>
+  requires std::same_as<EntityT, SemIR::Interface>
+static auto TryGetEntity(Context& context, SemIR::Inst inst)
+    -> const SemIR::EntityWithParamsBase* {
+  if (auto decl = inst.TryAs<SemIR::InterfaceDecl>()) {
+    return &context.interfaces().Get(decl->interface_id);
+  } else {
+    return nullptr;
+  }
+}
+
+template <typename EntityT>
+  requires std::same_as<EntityT, SemIR::NamedConstraint>
+static auto TryGetEntity(Context& context, SemIR::Inst inst)
+    -> const SemIR::EntityWithParamsBase* {
+  if (auto decl = inst.TryAs<SemIR::NamedConstraintDecl>()) {
+    return &context.named_constraints().Get(decl->named_constraint_id);
+  } else {
+    return nullptr;
+  }
+}
```

Comment:
Optional: I think you can make these into a conventional overload set with something like this:

```suggestion
template <typename T>
struct Tag {};

static auto TryGetEntity(Tag<SemIR::Interface>, Context& context, SemIR::Inst inst)
    -> const SemIR::EntityWithParamsBase* {
  if (auto decl = inst.TryAs<SemIR::InterfaceDecl>()) {
    return &context.interfaces().Get(decl->interface_id);
  } else {
    return nullptr;
  }
}

static auto TryGetEntity(Tag<SemIR::NamedConstraint>, Context& context, SemIR::Inst inst)
    -> const SemIR::EntityWithParamsBase* {
  if (auto decl = inst.TryAs<SemIR::NamedConstraintDecl>()) {
    return &context.named_constraints().Get(decl->named_constraint_id);
  } else {
    return nullptr;
  }
}
```

Then the callsite looks like `TryGetEntity(Tag<EntityT>{}, context, existing_decl_inst)`.

This technique trades off a slightly more awkward interface for a more straightforward implementation, so I'd be more hesitant to do it with a public function, but it can be a good fit for implementation details within a .cpp file.